src/transactions/test/SorobanTxTestUtils.cpp

```diff
@@ -571,6 +634,360 @@ ContractInvocationTest::createUploadWasmTx(TestAccount& source,
+}
+
+void
+AssetContractInvocationTest::transfer(TestAccount& fromAcc,
+                                      SCAddress const& toAddr, int64_t amount,
+                                      bool expectSuccess)
+{
+    SCVal toVal(SCV_ADDRESS);
+    toVal.address() = toAddr;
+
+    SCVal fromVal(SCV_ADDRESS);
+    fromVal.address() = makeAccountAddress(fromAcc.getPublicKey());
+
+    LedgerKey fromBalanceKey = makeBalanceKey(fromAcc.getPublicKey());
+    LedgerKey toBalanceKey = makeBalanceKey(toAddr);
+
+    auto fn = makeSymbol("transfer");
+    Operation transfer;
+    transfer.body.type(INVOKE_HOST_FUNCTION);
+    auto& ihf = transfer.body.invokeHostFunctionOp().hostFunction;
+    ihf.type(HOST_FUNCTION_TYPE_INVOKE_CONTRACT);
+    ihf.invokeContract().contractAddress = mContractID;
+    ihf.invokeContract().functionName = fn;
+    ihf.invokeContract().args = {fromVal, toVal, makeI128(amount)};
+
+    transfer.body.invokeHostFunctionOp().auth = {
+        getInvocationAuth(ihf.invokeContract())};
+
+    SorobanResources resources;
+    resources.instructions = 2'000'000;
+    resources.readBytes = 2000;
+    resources.writeBytes = 1072;
+
+    resources.footprint.readOnly = mContractKeys;
+
+    bool fromIsIssuer = false;
+    bool toIsIssuer = false;
+    if (mAsset.type() != ASSET_TYPE_NATIVE)
+    {
+        LedgerKey issuerLedgerKey(ACCOUNT);
+        issuerLedgerKey.account().accountID = getIssuer(mAsset);
+        resources.footprint.readOnly.emplace_back(issuerLedgerKey);
+
+        if (getIssuer(mAsset) == fromAcc.getPublicKey())
+        {
+            fromIsIssuer = true;
+        }
+        else
+        {
+            resources.footprint.readWrite.emplace_back(fromBalanceKey);
+        }
+
+        if (toAddr.type() == SC_ADDRESS_TYPE_ACCOUNT &&
+            getIssuer(mAsset) == toAddr.accountId())
+        {
+            toIsIssuer = true;
+        }
+        else
+        {
+            resources.footprint.readWrite.emplace_back(toBalanceKey);
+        }
+    }
+    else
+    {
+        resources.footprint.readWrite = {fromBalanceKey, toBalanceKey};
+    }
+
+    auto preTransferFromBalance = mAsset.type() == ASSET_TYPE_NATIVE
+                                      ? fromAcc.getBalance()
+                                      : fromAcc.getTrustlineBalance(mAsset);
+    auto preTransferToBalance = getBalance(toAddr);
+
+    // submit operation
+    auto tx = sorobanTransactionFrameFromOps(mApp->getNetworkID(), fromAcc,
+                                             {transfer}, {}, resources, 100,
+                                             DEFAULT_TEST_RESOURCE_FEE);
+    invokeTx(tx, expectSuccess, false);
+    if (expectSuccess)
+    {
+        auto postTransferFromBalance =
+            mAsset.type() == ASSET_TYPE_NATIVE
+                ? fromAcc.getBalance()
+                : fromAcc.getTrustlineBalance(mAsset);
+
+        auto postTransferToBalance = getBalance(toAddr);
+
+        REQUIRE((fromIsIssuer ||
```

Comment:
What is the expected balance when from/to is issuer?

Reply:
The issuer cannot hold a balance, so the amount gets burned or minted. I can add a check to verify that the contract balance ledger entry doesn't exist for the `to` address.

Reply:
Done

Reply:
Yeah, that's what I meant - if we expect no entries, then we need to make sure there are no entries. It would also make sense to have this check for `fromIsIssuer`, right? (to make sure we haven't created the issuer trustline.

Reply:
Done